game-app/game-core/src/main/java/games/strategy/triplea/odds/calculator/AggregateResults.java

```diff
@@ -107,103 +195,362 @@ public double getAverageTuvSwing(
-            .map(BattleResults::getRemainingDefendingUnits)
-            .mapToDouble(Collection::size)
-            .sum()
-        / results.size();
+  /**
+   * Returns the standard deviation of the TUV swing.
+   *
+   * <p>Note: The result of this function only makes sense for battle with the same initial setup,
+   * e.g. results coming from simulations of the same battle. The {@code attackers} and {@code
+   * defenders} are the units attacking and defending in this setup, respectively.
+   *
+   * <p>{@code attacker}, {@code defender} and {@code data} are used to derive the TUV values of the
+   * units.
+   *
+   * @param attacker the attacking player
+   * @param defender the defending player
+   * @param data the game data
+   * @return the standard deviation for the TUV swing of the battles.
+   */
+  public double getAverageTuvSwingStandardDeviation(
+      final GamePlayer attacker, final GamePlayer defender, final GameData data) {
+    // The variance is invariant under a constant shift.  Thus, the variance of
+    // the TUV of left-over units and the variance of the TUV of lost units is
+    // the same as they differ only be the (constant) total TUV the units at
+    // the start of the battle.
+    final IntegerMap<UnitType> attackerCostsForTuv = TuvUtils.getCostsForTuv(attacker, data);
+    final IntegerMap<UnitType> defenderCostsForTuv = TuvUtils.getCostsForTuv(defender, data);
+    // The 'true' argument means calculate the variance using bias correction.
+    final StandardDeviation stdDev = new StandardDeviation(true);
+    final double s =
+        stdDev.evaluate(
+            results.stream()
+                .mapToDouble(
+                    result ->
+                        TuvUtils.getTuv(result.getRemainingAttackingUnits(), attackerCostsForTuv)
+                            - TuvUtils.getTuv(
+                                result.getRemainingDefendingUnits(), defenderCostsForTuv))
+                .toArray());
+    return Double.isNaN(s) ? 0 : s;
   }
 
-  double getAverageDefendingUnitsLeftWhenDefenderWon() {
-    if (results.isEmpty()) {
-      return 0.0;
-    }
-    double count = 0;
-    double total = 0;
-    for (final BattleResults result : results) {
-      if (result.defenderWon()) {
-        count += result.getRemainingDefendingUnits().size();
-        total += 1;
-      }
-    }
-    if (total <= 0) {
-      return 0;
-    }
-    return count / total;
+  /**
+   * Returns the average number of attacking units surviving the battles.
+   *
+   * @return the average of surviving attacking units
+   */
+  public double getAverageAttackingUnitsLeft() {
+    final Mean mean = new Mean();
+    final double m =
+        mean.evaluate(
+            results.stream()
+                .map(BattleResults::getRemainingAttackingUnits)
+                .mapToDouble(Collection::size)
+                .toArray());
+    // If there are no battles, mean.evaluate() returns Double.NaN.
+    return Double.isNaN(m) ? 0 : m;
```

Comment:
I wonder if this can be made a bit easier to understand. As a suggestion, do you think something like this would be easier to follow (or just all in the eye of the beholder?)?
```
Double[] attackingUnitsLeft = results.stream()
                .map(BattleResults::getRemainingAttackingUnits)
                .mapToDouble(Collection::size)
                .toArray();
return attackingUnitsLeft.length == 0 ? 0 : new Mean(attackingUnitsLeft);
```
This avoids needing to use the `isNan` or comment that `mean` returns that value for empty arrays, it's more explicit in the above. We also reduce the number of intermediate variables as well. I suppose a similar pattern could apply to the other methods as well. WDYT?

Reply:
What I don't like at the suggested code is that we use the "check first" pattern, a pattern which is often discourage (in particular when doing I/O) better put the check afterwards which is summed up in the mantra "Better ask for forgiveness than permission". So, one should aim for the "check afterwards" pattern (if the library API allows for this), where the return value of the actual function is checked.  Then and only then, we can be sure that a function call worked/didn't work.

Applied to the above code, this means that Mean.evaluate() could still return NaN.  (But looking at some version of code of the library, I do not see any case besides no input data, where this could happen).
If we use the NaN check afterwards we are assured to have caught every possible outcome and never return NaN in our function.  However, the above is really nitpicking for such a simple API usage.

Personally, I like my version more (not a surprise) but I also have a reason for it:
In that version the (core of the) algorithm is complete in the line `double s = mean.evalute(...)`.  Then afterwards it follows the check for rare corner cases. In the proposed version these two things are mixed, first we do stuff to generate an array, then we check the corner cases, then we finish the calculation of the mean. If I'm not familiar with a code base, I'd like first to know what do we do there and then afterwards understand (or see) what other cases have additionally be taken care of (e.g. error handling, corner cases, ...).

I believe in the end it doesn't really matter both are reasonably fine.


After contemplating a little more.  I think we should even go a similar way as the library and return NaN in case we cannot calculate the requested quantity.  Do we really gain something by returning 0 if we cannot calculate the mean? Or would it better to return NaN to tell the caller "sorry, not valid what you trying to do".

I even wonder whether we should do the NaN check at all.  Do we really gain something by return 0 when the reality is "sorry we cannot calculate the mean/stddev".  In that case, NaN is a valid return value indicating "something's wrong with your input data". (Usually you would throw an exception, however, for the `getAverageAttackingUnitsLeftWhenAttackerWon()` the caller cannot reasonable ensure that there is a won battle.)

Deciding so would make the NaN checks redundant and the only change would be that the user gets a `NaN ± NaN / 8` for such cases, that however we can circumvent with a check when formatting that number and print `N/A` instead. (There is no other call to the `getAverageAttackingUnitsLeftWhenAttackerWon()` function.)
What do you think about this proposal?

Reply:
Good analysis. Returning and displaying `N/A` seems reasonable.